### DISCUSSION TITLE
[@types/react-table]: Column id property is always a string

### DISCUSSION THREAD

**Discussion by salemhilal** [DefinitelyTyped/DefinitelyTyped#58018] [Issues with a @types package]
The `Column` type uses a helper to define its ID property [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-table/index.d.ts#L74):
```ts
  //...
        | { Header: string; }
        | ({ id: IdType<D>; } 
        & { //...
```

That property uses these helper types:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/457855832440d90c5dddd0179b554123498c0b67/types/react-table/index.d.ts#L872-L873

As written, `IdType` always reduces to `string`. `StringKey` always returns a union of values that extend `string`, which `IdType` then unions with `string`. Since everything in that union type extends `string`, it seems to get collapsed. 

It seems like `IdType` tries to handle the case where `StringKey` might return `never`. A fix for this might be to check for `never` like this:

```ts
export type IdTypeTheSequel<D> = StringKey<D> extends never ? string : StringKey<D>;
```

[Here is a playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/PTAEFEA8EsGcBdoDsDmoDGB7AJgUwFC6QAOmATvKPAJ7G6gDK8ZyKA0rtQDwAiAfKAC8ESMwCG6eFwDWnTADNQPADSgELVHwDchEuUo06oAJLYAKrVy8BwphvadroAD5rmrHfhCgACpni4SIhiADag8tCQuqQUVJYm5pYATE627qgc3PygRAFI2LCgSLgAbrhkoAD8bvagAFyM6Q5Z2vheYAyYALb06qygstSw+MgBZPIS9MagAN74oAugYg1IAK5dAEblyvOLGw0bmJghuGJIO4sYDX2oF4sAjNdN+AC+bd4Acpg1-YPDo+UJuh6AApWa7BaPIrrLZkO4LJIHI4nM7w0AAZie9lebUM9DM9yECQsdC4xm0l28INWCFAAHIbig6fg8aAzEkiaYSVZyVpKWBqbSGU1mayzOiiXZWJkyRTvHSxHSXPSNkrXHT0KL4mYACySpoykFysDFMpkXHagCsnMSdBSvNA8sVyrpqpdmpZ2oAbDbuSkjXzvIygA) with demonstrations of the current problem and a possible fix.

[a comment by an automated account is not shown]

**Comment by ggascoigne** [comment]
I'll need to dig into this a bit more.  The playground is really useful, thank you.

That said, I've got a feeling that the issue is to do with support for nested keys.

IIRC column keys can have a dotted notation (which is useful, but really awkward to type), and so the ID can refer to the property of a nested object.  I *think* that this is why IdType isn't just StringKey but I'll have to go back through the change lots to see if I can work it out again :)

**Reply by salemhilal** [reply]
Thank you for your quick response! I'm admittedly a newer user of react-table, so I'm still getting a better picture of all the ways these types might be used.

**Reply by ggascoigne** [reply]
you might find https://github.com/ggascoigne/react-table-example useful if you are looking for a typescript example.  It's reasonably up to date.
